File: newsmodule/src/main/java/zjm/cst/dhu/newsmodule/utils/UrlImageGetter.java
package zjm.cst.dhu.newsmodule.utils;

import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.media.ThumbnailUtils;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.text.Html;
import android.widget.TextView;

import com.bumptech.glide.Glide;
import com.bumptech.glide.load.engine.DiskCacheStrategy;
import com.bumptech.glide.load.resource.drawable.GlideDrawable;
import com.bumptech.glide.request.RequestListener;
import com.bumptech.glide.request.animation.GlideAnimation;
import com.bumptech.glide.request.target.SimpleTarget;
import com.bumptech.glide.request.target.Target;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutionException;

import okhttp3.ResponseBody;
import retrofit2.Retrofit;
import rx.Subscriber;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.functions.Func1;
import rx.schedulers.Schedulers;
import zjm.cst.dhu.library.Constant;
import zjm.cst.dhu.library.MyApplication;
import zjm.cst.dhu.library.R;
import zjm.cst.dhu.library.api.repository.Repository;
import zjm.cst.dhu.library.utils.RetrofitUtil;

import static android.text.Html.FROM_HTML_MODE_COMPACT;
import static android.text.Html.FROM_HTML_MODE_LEGACY;
import static zjm.cst.dhu.library.Constant.NETEASE_NEWS_BASE_URL;

/**
 * Created by zjm on 2017/5/14.
 */

public class UrlImageGetter implements Html.ImageGetter {

    private Map<Integer, Drawable> drawableHashMap = new HashMap<>();

    private TextView mTextView;
    private String mBody;
    private Fragment fragment;
    private int imgCount;
    private int nowCount=0;
    private int mImgWidth;
    private int mImgHeight;

    public UrlImageGetter(TextView textView, String body, int imgCount,Fragment fragment) {
        this.mTextView = textView;
        this.mBody = body;
        this.fragment = fragment;
        this.imgCount=imgCount;
        mImgWidth = mTextView.getWidth();
    }

    @Override
    public Drawable getDrawable(final String source) {
        Drawable drawable = drawableHashMap.get(source.hashCode());
        if (drawable == null) {
            Glide.with(fragment).load(source)
                    .diskCacheStrategy(DiskCacheStrategy.ALL)
                    .into(new SimpleTarget<GlideDrawable>(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL) {
                        @Override
                        public void onResourceReady(GlideDrawable resource, GlideAnimation<? super GlideDrawable> glideAnimation) {
                            float imgWidth = resource.getIntrinsicWidth();
                            float imgHeight = resource.getIntrinsicHeight();
                            float rate = imgHeight / imgWidth;
                            mImgHeight = (int) (mImgWidth * rate);
                            resource.setBounds(0, 0, mImgWidth, mImgHeight);
                            drawableHashMap.put(source.hashCode(), resource);
                            nowCount++;
                            if(imgCount==nowCount){
                                loadImage();
                            }
                        }
                    });
            return createPicPlaceholder();
        } else {
            return drawable;
        }
    }

    private void loadImage() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.N) {
            mTextView.setText(Html.fromHtml(mBody,
                    FROM_HTML_MODE_LEGACY, UrlImageGetter.this, null));
        } else {
            mTextView.setText(Html.fromHtml(mBody, UrlImageGetter.this, null));
        }
    }

    @SuppressWarnings("deprecation")
    private Drawable createPicPlaceholder() {
        Drawable drawable;
        drawable = new ColorDrawable(MyApplication.getContext().getResources().getColor(R.color.colorGreyDark));
        drawable.setBounds(0, 0, mImgWidth, mImgWidth / 3);
        return drawable;
    }
}

File: newsmodule/src/main/kotlin/zjm.cst.dhu.newsmodule/utils/UrlImageGetter.kt
package zjm.cst.dhu.newsmodule.utils

import android.graphics.drawable.ColorDrawable
import android.graphics.drawable.Drawable
import android.os.Build
import android.support.v4.app.Fragment
import android.text.Html
import android.widget.TextView

import com.bumptech.glide.Glide
import com.bumptech.glide.load.engine.DiskCacheStrategy
import com.bumptech.glide.load.resource.drawable.GlideDrawable
import com.bumptech.glide.request.animation.GlideAnimation
import com.bumptech.glide.request.target.SimpleTarget
import com.bumptech.glide.request.target.Target

import zjm.cst.dhu.library.MyApplication
import zjm.cst.dhu.library.R

import android.text.Html.FROM_HTML_MODE_LEGACY

/**
 * Created by zjm on 2017/5/14.
 */

class UrlImageGetter(private val mTextView: TextView, private val mBody: String,
                     private val imgCount: Int, private val fragment: Fragment)
    : Html.ImageGetter {

    private val drawableHashMap = HashMap<Int, Drawable>()
    private var nowCount = 0
    private val mImgWidth: Int = mTextView.width
    private var mImgHeight: Int = 0

    override fun getDrawable(source: String): Drawable {
        val drawable = drawableHashMap[source.hashCode()]
        if (drawable == null) {
            loadGlide(source)
            return createPicPlaceholder()
        } else {
            return drawable
        }
    }

    private fun loadGlide(source: String) {
        Glide.with(fragment).load(source)
                .diskCacheStrategy(DiskCacheStrategy.ALL)
                .into(object : SimpleTarget<GlideDrawable>(Target.SIZE_ORIGINAL, Target.SIZE_ORIGINAL) {
                    override fun onResourceReady(resource: GlideDrawable,
                                                 glideAnimation: GlideAnimation<in GlideDrawable>) {
                        val imgWidth = resource.intrinsicWidth.toFloat()
                        val imgHeight = resource.intrinsicHeight.toFloat()
                        val rate = imgHeight / imgWidth
                        mImgHeight = (mImgWidth * rate).toInt()
                        resource.setBounds(0, 0, mImgWidth, mImgHeight)
                        drawableHashMap.put(source.hashCode(), resource)
                        nowCount++
                        if (imgCount == nowCount) {
                            loadImage()
                        }
                    }
                })
    }

    private fun loadImage() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            mTextView.text = Html.fromHtml(mBody,
                    FROM_HTML_MODE_LEGACY, this@UrlImageGetter, null)
        } else {
            mTextView.text = Html.fromHtml(mBody, this@UrlImageGetter, null)
        }
    }

    private fun createPicPlaceholder():
            Drawable {
        val drawable: Drawable
        drawable = ColorDrawable(MyApplication.context!!.resources.getColor(R.color.colorGreyDark))
        drawable.setBounds(0, 0, mImgWidth, mImgWidth / 3)
        return drawable
    }
}
